README.md

```diff
@@ -1371,6 +1371,364 @@ With all of our builtin logic integrated into the codebase, we can implement any
+            bytes[i] = uint8(byte)
+        }
+        return lambdaworks.FeltFromBeBytes(&bytes)
+    }
+    ```
+
+###### Calling the poseidon permutation function
+
+Now that we have our felt types defined we can move on to wrapping the poseidon permutation function. The `poseidon_permute_comp` from `starknet_crypto` receives a mutable state of three felts as an array. To reduce the complexity of our wrapper we will be receiving three felts in our C function.
+
+We will define the following function in our C header file:
+
+```C
+// Computes the poseidon hash permutation over a state of three felts
+void poseidon_permute(felt_t, felt_t, felt_t);
+```
+
+And we will implement it in the rust lib file, using the types and conversions we implemented earlier:
+
+```rust
+use starknet_crypto::{poseidon_permute_comp, FieldElement};
+
+#[no_mangle]
+extern "C" fn poseidon_permute(
+    first_state_felt: Bytes,
+    second_state_felt: Bytes,
+    third_state_felt: Bytes,
+) {
+    // Convert state from C representation to FieldElement
+    let mut state_array: [FieldElement; 3] = [
+        field_element_from_bytes(first_state_felt),
+        field_element_from_bytes(second_state_felt),
+        field_element_from_bytes(third_state_felt),
+    ];
+    // Call poseidon permute comp
+    poseidon_permute_comp(&mut state_array);
+    // Convert state from FieldElement back to C representation
+    bytes_from_field_element(state_array[0], first_state_felt);
+    bytes_from_field_element(state_array[1], second_state_felt);
+    bytes_from_field_element(state_array[2], third_state_felt);
+}
+```
+
+And with our lib ready, all that is left is to make a go wrapper with our felt conversion functions that calls the C function:
+
+```go
+func PoseidonPermuteComp(poseidon_state *[3]lambdaworks.Felt) {
+ state := *poseidon_state
+ // Convert args to c representation
+ first_state_felt := toC(state[0])
+ second_state_felt := toC(state[1])
+ third_state_felt := toC(state[2])
+
+ // Compute hash using starknet_crypto C wrapper
+ C.poseidon_permute(&first_state_felt[0], &second_state_felt[0], &third_state_felt[0])
+ // Convert result to Go representation
+ var new_poseidon_state = [3]lambdaworks.Felt{
+  fromC(first_state_felt),
+  fromC(second_state_felt),
+  fromC(third_state_felt),
+ }
+ // Update poseidon state
+ *poseidon_state = new_poseidon_state
+}
+```
+
+Now that we can call a simple poseidon permutation function we can start implementing our poseidon builtin runner!
+
+##### Implementing the PoseidonBuiltinRunner
+
+We will start by defining our `PoseidonBuiltinRunner` and adding it to our VM when creating a `CairoRunner`:
+
+It will contain it's base and a cache of values that we will use later to optimze our `DeduceMemoryCell` method. The included field indicates if a builtin is used by the program, is used in proof_mode, as all builtins have to be present by default, but for now we will always set the included field to true.
```

Comment:
```suggestion
It will contain it's base and a cache of values that we will use later to optimize our `DeduceMemoryCell` method. The included field indicates if a builtin is used by the program, is used in proof_mode, as all builtins have to be present by default, but for now we will always set the included field to true.
```